### PR TITLE
2696: Invert `@SubclassMappings` with `@InheritInverseConfiguration`.

### LINKED ISSUE #2696: Allow @InheritInverseConfiguration with @SubclassMappings
using `@InheritInverseConfiguration` results in an error, I had to explicitly write the `@SubclassMapping`s manually. It would be useful to be able to inherit them, either through the standard annotation or with an ad-hoc one.

_Originally posted by @pennal in https://github.com/mapstruct/mapstruct/issues/2695#issuecomment-998924888_

### PR DESCRIPTION
**Pull request by Zegveld**
`@InheritConfiguration` still not possible with `@SubclassMapping`'s, but `@InheritInverseConfiguration` is possible with `@SubclassMapping`'s.

Added a new error in case of identical source classes in the `@SubclassMapping`'s.
For normal `@SubclassMapping`'s this was easy to spot, but with the `@InheritInverseConfiguration` it was suddenly possible, and you would end up with unreachable code.
```java
if ( A instanceof B ) {
  // do mapping
} else if ( A instanceof B ) { // <--- will never be true
  // do different mapping //  <--- unreachable code
}
```

Fixes #2696

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java:95]: We never use the return type from `isValidUsage`. Should we do that and assign it to `isConsistent` if it is `false`?

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java:188]: Why don't we pass the `message` and `typeUtils` to this method instead of dragging them in the `SubclassMappingOptions`? It doesn't feel right to me to add those things to the model.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java:37]: Where is this used?

**Review by filiphr — CHANGES_REQUESTED** [review]
Thanks for your work on this @Zegveld. I've left some comments that are unclear to me.

Also it would be good to add a test case where you can override something in the `InheritInverseConfiguration` method.

e.g.  It should be possible to fix `ErroneousInverseSubclassMapper` by doing

```
@SubclassMapping( target = Bike.class, source = BikeDto.class )
@SubclassMapping( target = Car.class, source = CarDto.class )
Vehicle mapInverse(VehicleDto dto)
```

Of course we also need to be able to carry over an additional `SubclassMapping` that is not defined. Is it possible to be able to distinquish certain things like this? I think yes, because the `source` has to anyways be unique in the first method and thus all the inversed mappings will have a unique `target`.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java
Comment (Zegveld) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java:95]: it is used at another location:
https://github.com/mapstruct/mapstruct/pull/2708/files#diff-86e6eb61ca7c458ab2ea7274f6950ab596031e8a450771263d5c7492ed151862R198

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java
Comment (Zegveld) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java:188]: I thought of doing that, but then I had trouble with determining the correct annotation etc for the message to pop-up on. I'll see if I can think of a better solution for this issue.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java
Comment (Zegveld) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java:37]: inside isValidUsage

**Comment by Zegveld** [comment]
> Also it would be good to add a test case where you can override something in the `InheritInverseConfiguration` method.
> 
> e.g. It should be possible to fix `ErroneousInverseSubclassMapper` by doing
> 
> ```
> @SubclassMapping( target = Bike.class, source = BikeDto.class )
> @SubclassMapping( target = Car.class, source = CarDto.class )
> Vehicle mapInverse(VehicleDto dto)
> ```
> 
> Of course we also need to be able to carry over an additional `SubclassMapping` that is not defined. Is it possible to be able to distinquish certain things like this? I think yes, because the `source` has to anyways be unique in the first method and thus all the inversed mappings will have a unique `target`.

I did think about this, and came to the following conclusion: Inheriting the inverse mappings only makes sense if both source and target are already unique. The reason for this is that the following behavior does not make sense:
```
class ElectricBike extends Bike {
  String voltage;
}
class ManualBike extends Bike {
  int numberOfGearShifts;
}
abstract class Bike {
  String size;
}
class BikeDto {
  String voltage;
  int numberOfGearShifts;
  String size;
}
@Mapper(subclassExhaustiveStrategy = SubclassExhaustiveStrategy.RUNTIME_EXCEPTION)
class BikeMapper {
    @SubclassMapping( source = ElectricBike.class, target = BikeDto.class )
    @SubclassMapping( source = ManualBike.class, target = BikeDto.class )
    BikeDto map(Bike bike);

    @InheritInverseConfiguration
    @SubclassMapping( source = BikeDto.class, target = ManualBike.class )
    Bike mapInverse(BikeDto dto);
}
```
using the following bit of code you get different results:
```
BikeMapper bikeMapper = /*retrieveMapper*/;
ElectricBike eBike = /*createElectricBike*/;
Bike thereAndBackAgain = bikeMapper.mapInverse(bikeMapper.map(eBike));
/* you would expect thereAndBackAgain to be the same eBike as it was before, 
but now it suddenly became a ManualBike. */
```
In this case you would want to have an `inverseCondition` or something, that way you can specify what to do instead of the instanceof check. Might also be useful for a normal `Condition` to replace the instanceof check and do the mapping still. This would even allow multiple `source`'s at that point, thus removing the uniqueness in case a condition is present. I believe this is better solved in a different PR.

Perhaps for now would be a better option to add a flag to the `InheritInverseConfiguration` with which you can disable the `SubclassMapping`s inheritence. That way the user can still use all other mappings reversed, but define all the `SubclassMapping`s manually.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java
Comment (Zegveld) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java:163]: Solved this by moving the SubclassValidator further outside. Now the same validator will be used for both the methods `SubclassMapping`s as well as the inverse inherited `SubclassMapping`s.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java
Comment (Zegveld) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java]: fixed this as specified using the `bitwise and` (`&` instead of `&&`) operator.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java
Comment (Zegveld) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java]: got removed during fixing of previous remark.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/SubclassMappingOptions.java:89]: I personally always have problems with the bitwise operator. Can we use something else here perhaps?

e.g.
```
return subclassValidator.isValidUsage( method, gem.mirror(), targetSubclass ) && isConsistent;
```

or even a simple if

```
if (!subclassValidator.isValidUsage( method, gem.mirror(), targetSubclass )) {
    isConsistent = false;
}
```

**Review by filiphr — COMMENTED** [review]
Thanks for your changes @Zegveld.

>I did think about this, and came to the following conclusion: Inheriting the inverse mappings only makes sense if both source and target are already unique. The reason for this is that the following behavior does not make sense:

I agree with you that it does not make sense to do that. However, it is up to you to do that, perhaps you have 10 different `SubclassMappings` and 9 out of those are unique, the 10th one is a special one reusing something else due to different reasons, it can be a catch all.

Now if you want to inherit those options your only option is to copy the 9 mappings and add a 10th specific one.

In my opinion we should take the same spirit as the `@Mapping`. You can map from the same source multiple times, when you reverse we allow people to provide different `@Mapping` for some target.

I do understand the change that you pulled the usage of the `SubclassValidator` further up, but I am not really feeling it. There are a lot of places that needed to be changed so we can use this. I will try to check your branch out and see if I can think of something about that

**Comment by Zegveld** [comment]
There is one problem with that, would the inherited mappings come before or after the mappings directly on the method?
Or perhaps they would come in between?
```
@SubclassMapping(target = before.class, ...)
@InheritInverseConfiguration
@SubclassMapping(target = after.class, ...)
```
Because the position is more important for subclassmappings than for normal mappings.

**Comment by filiphr** [comment]
The order of the `@SubclassMapping` is indeed an important one. If you ask me I would say that the order should be in the same place as it was before, unless it is a new not defined target, then it goes to the top (highest priority).

We can wait for feedback from users about this and perhaps even add support for custom ordering through something like `SubclassMapping#dependsOn`, but I would decouple that from this PR. This should behave the same way as it does for `@Mapping`

**Comment by Zegveld** [comment]
> The order of the `@SubclassMapping` is indeed an important one. If you ask me I would say that the order should be in the same place as it was before, unless it is a new not defined target, then it goes to the top (highest priority).

Actually it was easier to make the new order primary, and have any inverse inherited mappings that are redefined skipped.
This will allow you to redefine the entire order if needed, by writing it out. This way you still do not need to rewrite your normal `@Mapping`s just to change the `@SubclassMapping`s order.

for example:
```
@SubclassMapping( source = SourceA.class, target = TargetA.class )
@SubclassMapping( source = SourceB.class, target = TargetB.class )
@SubclassMapping( source = SourceC.class, target = TargetC.class )
@SubclassMapping( source = SourceD.class, target = TargetD.class )
@SubclassMapping( source = SourceA2.class, target = TargetA.class )
Target from(Source source);

@InheritInverseMapping
@SubclassMapping( source = TargetD.class, target = SourceD.class )
@SubclassMapping( source = TargetB.class, target = SourceB.class )
Source from(Target source);
```
will end up as
```
@SubclassMapping( source = TargetD.class, target = SourceD.class )
@SubclassMapping( source = TargetB.class, target = SourceB.class )
@SubclassMapping( source = TargetA.class, target = SourceA.class )
@SubclassMapping( source = TargetC.class, target = SourceC.class )
Source from(Target source);
```

Also added support for overriding a double target issue, by redefining it as seen in above example.

**Comment by filiphr** [comment]
>This way you still do not need to rewrite your normal `@Mappings` just to change the `@SubclassMappings` order.

I thought that this was implied already 😄 , as to me those are 2 independent things. Anyways, I am going through look at your last changes now

File: processor/src/test/java/org/mapstruct/ap/test/subclassmapping/SubclassMappingTest.java
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/subclassmapping/SubclassMappingTest.java:109]: Don't we want to test the inverse mapping here and check the type of the generated objects?

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java:53]: We can't use `TypeMirror` as a key in a Map. The `equals` implementation of `TypeMirror` might return `false` for types that are actually the same. That's why we use `Types#isSameType` to check these things. Can we not use the FQN of the type instead?

**Review by filiphr — CHANGES_REQUESTED** [review]
Went through it again and there are still some things that we need to sort out before we merge this.

Why do we need to the map for tracking the order? Why can't we do something like we do for the `mappingOptions`? I know that the `MappingOptions` have `equals` and `hashCode` implemented, but the `SubclassMappingOptions` do not have that.

**Comment by Zegveld** [comment]
> Went through it again and there are still some things that we need to sort out before we merge this.
> 
> Why do we need to the map for tracking the order? Why can't we do something like we do for the `mappingOptions`? I know that the `MappingOptions` have `equals` and `hashCode` implemented, but the `SubclassMappingOptions` do not have that.

Indeed we do not need the map. Originally I had the map because that one does replace it, while a Set does not replace the same item. But when rewriting I ended up checking if it is already there, and if so skipping it. Which basically becomes the behavior of the Set, so I should refactor this back into a Set and add the equals/hashcode to the SubclassMappingOptions.
This should also fix 2 of the issues with the code you mentioned.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java
Comment (Zegveld) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java]: Map not needed anymore, implemented an equals on SubclassMappingOptions that uses TypeUtils.isSameType with source MirrorType.

File: processor/src/test/java/org/mapstruct/ap/test/subclassmapping/SubclassMappingTest.java
Comment (Zegveld) [reply processor/src/test/java/org/mapstruct/ap/test/subclassmapping/SubclassMappingTest.java]: Indeed, fixed the tests.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java
Comment (Zegveld) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java]: It is no longer a Map, but a Set so it is now in sync again without creating a new HashSet.

File: processor/src/test/java/org/mapstruct/ap/test/subclassmapping/SubclassMappingTest.java
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/subclassmapping/SubclassMappingTest.java:94]: Don't we care only about the return type? I thought that this is what we are testing? At least that is what we test in the original method

**Review by filiphr — COMMENTED** [review]
Thanks a lot for your changes @Zegveld. I've added 2 more questions, one is about the testing, since I am confused a bit about it, and the second one is a small style question for a setter

File: processor/src/test/java/org/mapstruct/ap/test/subclassmapping/SubclassMappingTest.java
Comment (Zegveld) [reply processor/src/test/java/org/mapstruct/ap/test/subclassmapping/SubclassMappingTest.java:94]: That's is already covered in the previous test (line 57). This one is that the `@Mapping` is also inherited by the inverse subclass mapping.

Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/subclassmapping/SubclassMappingTest.java:94]: You are right, my mistake

**Review by filiphr — APPROVED** [review]
Thanks a lot for all the changes @Zegveld. It looks good to me now. Please go ahead with squashing and merging it.

**Pull request merged**
